server/internal/core/application/indexer.go

```diff
@@ -0,0 +1,553 @@
+	allVtxos, err := i.repoManager.Vtxos().GetAll(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	spendable := make([]domain.Vtxo, 0)
+	spent := make([]domain.Vtxo, 0)
+
+	for _, vtxo := range allVtxos {
+		if vtxo.Spent || vtxo.Swept {
+			spent = append(spent, vtxo)
+		} else {
+			spendable = append(spendable, vtxo)
+		}
+	}
+
+	txs, err := vtxosToTxs(spendable, spent)
+	if err != nil {
+		return nil, err
+	}
+
+	txsPaged, pageResp := paginate(txs, req.Page, maxPageSizeTxHistory)
+
+	return &TxHistoryResp{
+		Records:    txsPaged,
+		Pagination: pageResp,
+	}, nil
+}
+
+type vtxoKeyWithCreatedAt struct {
+	domain.VtxoKey
+	CreatedAt int64
+}
+
+func (i *indexerService) GetVtxoChain(ctx context.Context, req VtxoChainReq) (*VtxoChainResp, error) {
+	chainMap := make(map[vtxoKeyWithCreatedAt][]string)
+
+	outpoint := domain.VtxoKey{
+		Txid: req.VtxoKey.Txid,
+		VOut: req.VtxoKey.Vout,
+	}
+
+	if err := i.buildChain(ctx, outpoint, chainMap, true); err != nil {
+		return nil, err
+	}
+
+	chainSlice := make([]vtxoKeyWithCreatedAt, 0, len(chainMap))
+	for vtxo := range chainMap {
+		chainSlice = append(chainSlice, vtxo)
+	}
+
+	sort.Slice(chainSlice, func(i, j int) bool {
+		return chainSlice[i].CreatedAt > chainSlice[j].CreatedAt
+	})
+
+	pagedChainSlice, pageResp := paginate(chainSlice, req.Page, maxPageSizeVtxoChain)
+
+	txMap := make(map[string][]string)
+	for _, vtxo := range pagedChainSlice {
+		txMap[vtxo.Txid] = chainMap[vtxo]
+	}
+
+	return &VtxoChainResp{
+		Transactions: txMap,
+		Page:         pageResp,
+	}, nil
+}
+
+func (i *indexerService) buildChain(
+	ctx context.Context,
+	outpoint domain.VtxoKey,
+	chain map[vtxoKeyWithCreatedAt][]string,
+	isFirst bool,
+) error {
+	vtxos, err := i.repoManager.Vtxos().GetVtxos(ctx, []domain.VtxoKey{outpoint})
+	if err != nil {
+		return err
+	}
+
+	if isFirst && len(vtxos) == 0 {
+		return fmt.Errorf("vtxo not found for outpoint: %v", outpoint)
+	}
+
+	vtxo := vtxos[0]
+	key := vtxoKeyWithCreatedAt{
+		VtxoKey:   outpoint,
+		CreatedAt: vtxo.CreatedAt,
+	}
+	if _, ok := chain[key]; !ok {
+		chain[key] = make([]string, 0)
+	} else {
+		return nil
+	}
+
+	//finish chain if this is the leaf Vtxo
+	if !vtxo.IsPending() {
+		chain[key] = append(chain[key], vtxo.RoundTxid)
+		return nil
+	}
+
+	redeemPsbt, err := psbt.NewFromRawBytes(strings.NewReader(vtxo.RedeemTx), true)
+	if err != nil {
+		return err
+	}
+
+	for _, in := range redeemPsbt.UnsignedTx.TxIn {
+		chain[key] = append(chain[key], in.PreviousOutPoint.Hash.String())
+		parentOutpoint := domain.VtxoKey{
+			Txid: in.PreviousOutPoint.Hash.String(),
+			VOut: in.PreviousOutPoint.Index,
+		}
+
+		if err := i.buildChain(ctx, parentOutpoint, chain, false); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (i *indexerService) GetVirtualTxs(ctx context.Context, req VirtualTxsReq) (*VirtualTxsResp, error) {
+	vtxos, err := i.repoManager.Vtxos().GetAll(ctx)
```

Comment:
let's make use of the newly defined API `GetTxsWithTxids`. The impl of this API is easy in sqlite, while it's not that straight forward for badger... for the latter, I'm ok with returning an error `not implemented` as it's meant only for testing purposes

Reply:
Done